### PR TITLE
Added Flow, Test, TestTemplate classes

### PR DESCRIPTION
**Pull request by RomanShcherba**

### CODE REVIEW

**Comment by RlNolino** [comment]
First - names of your classes general - it doesn't give me any info what feature we are validating
Second - don't forgot about summary for methods, classes etc

File: PortSpeedTestTemplate/Flow.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+using System.Threading.Tasks;
+
+namespace PortSpeedTestTemplate
+{
+    public class Flow
+    {
+        private static readonly ILog log = LogManager.GetLogger(typeof(Flow));
+
+        private TestTemplate testTemplate;
+        private TestTemplate parameter;
```
Comment (RlNolino) [inline PortSpeedTestTemplate/Flow.cs]: Lets remove it, and instead it, we will have inharitance.

File: PortSpeedTestTemplate/Flow.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+        public void Prepare()
```
Comment (RlNolino) [inline PortSpeedTestTemplate/Flow.cs]: I should be bool, not void

File: PortSpeedTestTemplate/Flow.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            log.Info("Preparing test");
+            testTemplate.Initialize();
+        }
+        public void Run()
+        {
+            log.Info("Running test");
+        }
+        public void Clean()
```
Comment (RlNolino) [inline PortSpeedTestTemplate/Flow.cs]: I should be bool, not void

File: PortSpeedTestTemplate/Test.cs [hunk 4]
```diff
@@ -10,9 +11,19 @@ namespace PortSpeedTestTemplate
...
+        public Test(string testName, string description, TestTemplate parameter)
+        {
+            Name = testName;
+            Description = description;
+            Parameters = parameter;
         }
+
+        public string TestName { get; }
+
     }
```
Comment (RlNolino) [inline PortSpeedTestTemplate/Test.cs]: what diffrence between Name and TestName?

File: PortSpeedTestTemplate/TestTemplate.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            test.Add(new Test("Test 8", "Testing network card with 8 port, with speed 90 Gbps", parameter8));
+
+            var parameter9 = new TestTemplate(8, new Dictionary<int, int> { { 1, 10 }, { 2, 10 }, { 3, 0 }, { 4, 0 }, { 5, 25 }, { 6, 10 }, { 7, 10 }, { 8, 0 }, });
+            test.Add(new Test("Test 9", "Testing network card with 8 port, with speed 65 Gbps", parameter9));
+
+            var parameter10 = new TestTemplate(8, new Dictionary<int, int> { { 1, 10 }, { 2, 10 }, { 3, 10 }, { 4, 10 }, { 5, 10 }, { 6, 25 }, { 7, 10 }, { 8, 10 }, });
+            test.Add(new Test("Test 10", "Testing network card with 8 port, with speed 95 Gbps", parameter10));
+            return test;
+
+        }
```
Comment (RlNolino) [inline PortSpeedTestTemplate/TestTemplate.cs]: You didn't listen what I told you. Each of these parameters should be in it's own test class. You shouldn't have collection of data inside template, only if that data is common for all tests.

File: PortSpeedTestTemplate/Flow.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        private TestTemplate parameter;
+
+        public void Prepare()
+        {
+            log.Info("Preparing test");
+            testTemplate.Initialize();
+        }
+        public void Run()
+        {
+            log.Info("Running test");
```
Comment (RlNolino) [inline PortSpeedTestTemplate/Flow.cs]: What exactly is here happening? I see only that it "running tests" that doesn't give any info. Do we set something for test? What steps are we doing here?

File: PortSpeedTestTemplate/Flow.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        }
+        public void Clean()
+        {
+            log.Info("Cleaning up after test");
+            testTemplate.TearDown();
```
Comment (RlNolino) [inline PortSpeedTestTemplate/Flow.cs]: Same, Clean make cleaning - it's about nothing. Waht steps inside?

**Comment by RlNolino** [comment]
I don't see any method. Current version of tests looks like - testing. What we are testing, how etc. no such info. Just testing.
This hapanns cause you dont build steps for your test case.

File: PortSpeedTestTemplate/WebCardSpeedTestTemplate.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+            PortSpeed = portSpeed;
+        }
+
+        public void Initialize()
+        {
+            log.Info("Initialize test");
+        }
+        public void SetPortSpeed(int portNumber, int portSeed)
+        {
+            log.Info($"Setting port speed {portNumber} to {portSeed}");
```
Comment (RlNolino) [inline PortSpeedTestTemplate/WebCardSpeedTestTemplate.cs]: What is portSeed?

File: PortSpeedTestTemplate/Program.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+        static void Main()
+        {
+            WebCardSpeedTest webCardSpeedTest = new WebCardSpeedTest(0, "Test", "Description", new WebCardSpeedTestTemplate(1, new Dictionary<int, int>()));
+            var tests = webCardSpeedTest.TestData();
+
+            foreach (var test in tests)
+            {
+                test.DisplayInfo();
+                test.Execute();
+                Console.WriteLine();
```
Comment (RlNolino) [inline PortSpeedTestTemplate/Program.cs]: Why it's here? And why you need it?

File: .gitignore [hunk 10]
```diff
@@ -360,4 +360,9 @@ MigrationBackup/
...
 
 # Fody - auto-generated XML schema
-FodyWeavers.xsd
\ No newline at end of file
+FodyWeavers.xsd
+/PortSpeedTestTemplate/Flow.cs
+/PortSpeedTestTemplate/Flow.cs
+/PortSpeedTestTemplate/Test.cs
+/PortSpeedTestTemplate/TestTemplate.cs
+/PortSpeedTestTemplate/TestTemplate.cs
```
Comment (RlNolino) [inline .gitignore]: It shouldn't be here

File: PortSpeedTestTemplate/WebCardSpeedTestFlow.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        public void Execute()
+        {
+            log.Info($"Executing test {TestName} - {Description}");
+            WebCardSpeedTestFlow flow = new WebCardSpeedTestFlow(parameters);
+            flow.Prepare();
+            flow.Run();
+            flow.Clean();
+        }
```
Comment (RlNolino) [inline PortSpeedTestTemplate/WebCardSpeedTestFlow.cs]: why it's here?

File: PortSpeedTestTemplate/WebCardSpeedTest.cs [hunk 12]
```diff
@@ -0,0 +1,70 @@
+﻿using log4net;
+using System.Linq;
+
+namespace PortSpeedTestTemplate
+{
+    public class WebCardSpeedTest : WebCardSpeedTestFlow
+    {
+        public static readonly ILog log = LogManager.GetLogger(typeof(WebCardSpeedTest));
```
Comment (RlNolino) [inline PortSpeedTestTemplate/WebCardSpeedTest.cs]: What is it?

File: PortSpeedTestTemplate/WebCardSpeedTestFlow.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
+﻿
+using log4net;
+
+namespace PortSpeedTestTemplate
+{
+    public class WebCardSpeedTestFlow : WebCardSpeedTestTemplate
+    {
+        private static readonly ILog log = LogManager.GetLogger(typeof(WebCardSpeedTest));
```
Comment (RlNolino) [inline PortSpeedTestTemplate/WebCardSpeedTestFlow.cs]: what it's doing?

File: PortSpeedTestTemplate/NetworkCardSpeedTest.cs [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+        public const string Name = "WebCardSpeedTest";
+
+        /// <summary>
+        /// Description of test
+        /// </summary>
+        public const string Description = "Test with mode";
+
+        #endregion
+
+        #region Data set
```
Comment (RlNolino) [inline PortSpeedTestTemplate/NetworkCardSpeedTest.cs]: The region is about type of members inside, and not about it's name. It should be Properties

File: PortSpeedTestTemplate/NetworkCardSpeedTest.cs [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        /// <summary>
+        /// Tuple of dataset
```
Comment (RlNolino) [inline PortSpeedTestTemplate/NetworkCardSpeedTest.cs]: Why you are writing that it is turple? We can see type in diffrent way, no need. More info about data that should be here

File: PortSpeedTestTemplate/NetworkCardSpeedTestFlow.cs [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    /// <summary>
+    /// Network card speed test flow 
+    /// </summary>
+    public class NetworkCardSpeedTestFlow : NetworkCardSpeedTestTemplate
+    {
+        #region Execution methods
+        /// <summary>
+        /// Prepare for test
+        /// </summary>
+        /// <returns></returns>
```
Comment (RlNolino) [inline PortSpeedTestTemplate/NetworkCardSpeedTestFlow.cs:15]: Do't need to fix, but you should also write type that the method will return

File: PortSpeedTestTemplate/NetworkCardSpeedTestFlow.cs [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+        /// <summary>
+        /// Clean after test
+        /// </summary>
+        /// <returns></returns>
+        public override bool Clean()
+        {
+            RestoreDefaultMode();
+            RestoreDefaultSpeed();
+            Console.WriteLine("\nCleaning up after test");
```
Comment (RlNolino) [inline PortSpeedTestTemplate/NetworkCardSpeedTestFlow.cs]: Why it a the end of method?

File: PortSpeedTestTemplate/NetworkCardSpeedTestFlow.cs [hunk 18]
```diff
@@ -0,0 +1,63 @@
+﻿//Company's property 2024
+
+namespace PortSpeedTestTemplate
+{
+    /// <summary>
+    /// Network card speed test flow 
+    /// </summary>
+    public class NetworkCardSpeedTestFlow : NetworkCardSpeedTestTemplate
+    {
+        #region Execution methods
```
Comment (RlNolino) [inline PortSpeedTestTemplate/NetworkCardSpeedTestFlow.cs]: There can be Public methods region or Private or Protected or Private and Protected methods region - please rename

File: PortSpeedTestTemplate/NetworkCardSpeedTestFlow.cs [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+        public override bool Clean()
+        {
+            RestoreDefaultMode();
+            RestoreDefaultSpeed();
+            Console.WriteLine("\nCleaning up after test");
+            return true;
+        }
+        #endregion
+
+        #region Runner
```
Comment (RlNolino) [inline PortSpeedTestTemplate/NetworkCardSpeedTestFlow.cs]: Public methods

File: PortSpeedTestTemplate/NetworkCardSpeedTestTemplate.cs [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+
+namespace PortSpeedTestTemplate
+{
+    /// <summary>
+    /// Network card speed test template
+    /// </summary>
+    public abstract class NetworkCardSpeedTestTemplate : Program
+    {
+
+        #region Execution methods
```
Comment (RlNolino) [inline PortSpeedTestTemplate/NetworkCardSpeedTestTemplate.cs]: Wrong name for region

File: PortSpeedTestTemplate/NetworkCardSpeedTestTemplate.cs [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+        /// <param name="portNumber">
+        /// Number of port
+        /// </param>
+        public void CheckPortStatus(int portNumber)
+        {
+            Console.WriteLine($"Checking status of port{portNumber}");
+        }
+        #endregion
+        #region Abstract flow methods
+        /// <summary>
```
Comment (RlNolino) [inline PortSpeedTestTemplate/NetworkCardSpeedTestTemplate.cs]: Add lines before and after regions and also wrong name

**Review by RlNolino — CHANGES_REQUESTED** [review]

File: PortSpeedTestTemplate/NetworkCardSpeedTestTemplate.cs (hunk 21 above)
Comment (RomanShcherba) [reply PortSpeedTestTemplate/NetworkCardSpeedTestTemplate.cs]: Done

File: PortSpeedTestTemplate/NetworkCardSpeedTestTemplate.cs (hunk 20 above)
Comment (RomanShcherba) [reply PortSpeedTestTemplate/NetworkCardSpeedTestTemplate.cs]: Region renamed

File: PortSpeedTestTemplate/NetworkCardSpeedTestFlow.cs (hunk 19 above)
Comment (RomanShcherba) [reply PortSpeedTestTemplate/NetworkCardSpeedTestFlow.cs]: Done

File: PortSpeedTestTemplate/NetworkCardSpeedTestTemplate.cs [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+
+        #region Protected methods
+
+        /// <summary>
+        /// Active card mode
+        /// </summary>
+        /// <param name="mode">
+        /// Card mode
+        /// </param>
+        protected static void CheckActiveCardMode(byte mode)
```
Comment (RlNolino) [inline PortSpeedTestTemplate/NetworkCardSpeedTestTemplate.cs:48]: I miss it, so leave it like it is. But for the future your methods will be not static and should be bool

**Review by RlNolino — APPROVED** [review]

**Pull request merged**